File: backend/src/ips_app/ports/driving/http/feature.py
from abc import ABC, abstractmethod
from typing import Optional, List, Tuple, Any
from ips_app.domain.models.feature import Feature
from ips_app.domain.models.permission import Permission

class FeatureHTTPPort(ABC):
    @abstractmethod
    async def add_feature(self, name: str, description: str) -> Feature:
        """Add a new feature."""
        ...

    @abstractmethod
    async def get_feature(self, feature_id: Any) -> Feature:
        """Get a feature by its ID."""
        ...

    @abstractmethod
    async def get_features(
        self, 
        page: int, 
        limit: int, 
        cursor_id: Optional[Any] = None, 
        search: Optional[str] = None
    ) -> Tuple[List[Feature], int]:
        """Get a list of features with pagination."""
        ...

    @abstractmethod
    async def set_feature(
        self, 
        feature_id: Any, 
        name: Optional[str] = None, 
        description: Optional[str] = None
    ) -> Feature:
        """Update a feature's basic information."""
        ...

    @abstractmethod
    async def set_feature_preferences(self, feature_id: Any, preferences: bytes) -> Feature:
        """Update a feature's preferences."""
        ...

    @abstractmethod
    async def remove_feature(self, feature_id: Any) -> str:
        """Remove a feature."""
        ...

    @abstractmethod
    async def add_permissions_to_feature(self, feature_id: Any, permission_ids: List[Any]) -> Feature:
        """Bind permissions to a feature."""
        ...

    @abstractmethod
    async def remove_permissions_from_feature(self, feature_id: Any, permission_ids: List[Any]) -> Feature:
        """Unbind permissions from a feature."""
        ...

    @abstractmethod
    async def get_feature_permissions(self, feature_id: Any) -> List[Permission]:
        """Get permissions bound to a feature."""
        ...

    @abstractmethod
    async def get_accessible_features(self, user_id: Any) -> List[Feature]:
        """Get all features accessible by a user."""
        ...

    @abstractmethod
    async def can_access_feature(self, user_id: Any, feature_id: Any) -> bool:
        """Check if a user can access a specific feature."""
        ...

File: backend/internal/ports/driving/http/feature.go
package portsdrivinghttp

import (
	"context"

	domainmodels "github.com/uwb-ips-dteti-ugm/ips-app/backend/internal/domain/models"
)

type Feature interface {
	// Features
	AddFeature(ctx context.Context, name string, description string) (feature domainmodels.Feature, err error)
	GetFeature(ctx context.Context, featureId int64) (feature domainmodels.Feature, err error)
	GetFeatures(ctx context.Context, page int, limit int, cursorId *int64, search *string) (features []domainmodels.Feature, total int, err error)
	SetFeature(ctx context.Context, featureId int64, name *string, description *string) (feature domainmodels.Feature, err error)
	SetFeaturePreferences(ctx context.Context, featureId int64, preferences []byte) (feature domainmodels.Feature, err error)
	RemoveFeature(ctx context.Context, featureId int64) (message string, err error)

	// Feature-Permission bindings
	AddPermissionsToFeature(ctx context.Context, featureId int64, permissionIds []int64) (feature domainmodels.Feature, err error)
	RemovePermissionsFromFeature(ctx context.Context, featureId int64, permissionIds []int64) (feature domainmodels.Feature, err error)
	GetFeaturePermissions(ctx context.Context, featureId int64) (permissions []domainmodels.Permission, err error)

	// Feature access
	GetAccessibleFeatures(ctx context.Context, userId int64) (features []domainmodels.Feature, err error)
	CanAccessFeature(ctx context.Context, userId int64, featureId int64) (canAccess bool, err error)
}
